libwallet/assets/btc/dex-wallet.go

```diff
@@ -175,90 +362,550 @@ func (dw *DEXWallet) SignTx(tx *wire.MsgTx) error {
+		// accuracy of confs is not significant. When it's our output, we'll
+		// know the block and won't end up here. (even if we did, we just end up
+		// sending out some inaccurate Data-severity notifications to the UI
+		// until the match progresses)
+		return confs, true, nil
+	}
+
+	// unspent
+	return confs, false, nil
+}
+
+// confirmations looks for the confirmation count and spend status on a
+// transaction output that pays to this wallet.
+func (dw *DEXWallet) confirmations(txHash *chainhash.Hash, vout uint32) (blockHash *chainhash.Hash, confs uint32, spent bool, err error) {
+	details, err := dw.walletTransaction(txHash)
+	if err != nil {
+		return nil, 0, false, err
+	}
+
+	if details.Block.Hash != (chainhash.Hash{}) {
+		blockHash = &details.Block.Hash
+		height, err := dw.GetChainHeight()
+		if err != nil {
+			return nil, 0, false, err
+		}
+		confs = uint32(confirms(details.Block.Height, height))
+	}
+
+	spent, found := outputSpendStatus(details, vout)
+	if found {
+		return blockHash, confs, spent, nil
+	}
+
+	return blockHash, confs, false, dexbtc.SpentStatusUnknown
+}
+
+// Part of dexbtc.Wallet interface.
+func (dw *DEXWallet) Locked() bool {
+	return dw.w.Locked()
+}
+
+// Part of dexbtc.Wallet interface.
+func (dw *DEXWallet) WalletLock() error {
+	dw.w.Lock()
+	return nil
+}
+
+// Part of dexbtc.Wallet interface.
+func (dw *DEXWallet) WalletUnlock(pw []byte) error {
+	return dw.w.Unlock(pw, nil)
+}
+
+// GetBlockHeader gets the *dexbtc.BlockHeader for the specified block hash. It also
+// returns a bool value to indicate whether this block is a part of main chain.
+// For orphaned blocks header.Confirmations is negative.
+// Part of dexbtc.TipRedemptionWallet interface.
+func (dw *DEXWallet) GetBlockHeader(blockHash *chainhash.Hash) (header *dexbtc.BlockHeader, mainchain bool, err error) {
+	hdr, err := dw.cl.CS.GetBlockHeader(blockHash)
+	if err != nil {
+		return nil, false, err
+	}
+
+	tip, err := dw.cl.CS.BestBlock()
+	if err != nil {
+		return nil, false, fmt.Errorf("BestBlock error: %v", err)
+	}
+
+	blockHeight, err := dw.cl.CS.GetBlockHeight(blockHash)
+	if err != nil {
+		return nil, false, err
+	}
+
+	confirmations := int64(-1)
+	mainchain = dw.blockIsMainchain(blockHash, blockHeight)
+	if mainchain {
+		confirmations = int64(confirms(blockHeight, tip.Height))
+	}
+
+	return &dexbtc.BlockHeader{
+		Hash:              hdr.BlockHash().String(),
+		Confirmations:     confirmations,
+		Height:            int64(blockHeight),
+		Time:              hdr.Timestamp.Unix(),
+		PreviousBlockHash: hdr.PrevBlock.String(),
+	}, mainchain, nil
+}
+
+// Part of dexbtc.Wallet interface.
+func (dw *DEXWallet) GetBestBlockHeader() (*dexbtc.BlockHeader, error) {
+	hash, err := dw.GetBestBlockHash()
+	if err != nil {
+		return nil, err
+	}
+	hdr, _, err := dw.GetBlockHeader(hash)
+	return hdr, err
+}
+
+// Part of dexbtc.Wallet interface.
+func (dw *DEXWallet) Connect(_ context.Context, _ *sync.WaitGroup) (err error) {
+	return nil
```

Comment:
Can emulate the upstream connect method a bit here: start a goroutine to periodically clean the block cache and exit the goroutine when the ctx is canceled.

Reply:
yh.